javascript/packages/cli/src/cli.ts

```diff
@@ -49,52 +48,81 @@ const downloadBinaries = async (binaries: string[]): Promise<void> => {
+          decorators.cyan("Binary name:") +
+          " {filename}",
+      },
+      cliProgress.Presets.shades_grey,
+    );
+
     for (let binary of binaries) {
       promises.push(
-        new Promise<void>(async (resolve) => {
-          const { url, name } = options[binary];
-          const { data, headers } = await axios({
-            url,
-            method: "GET",
-            responseType: "stream",
-          });
-          const totalLength = headers["content-length"];
-
-          const progressBar = new progress(
-            "-> downloading [:bar] :percent :etas",
-            {
-              width: 40,
-              complete: "=",
-              incomplete: " ",
-              renderThrottle: 1,
-              total: parseInt(totalLength),
-            },
-          );
+        new Promise<void>(async (resolve, reject) => {
+          let result = options[binary];
+          if (!result) {
+            console.log("options", options, "binary", binary);
+            throw new Error("Binary is not defined");
+          }
+          const { url, name } = result;
 
-          const writer = fs.createWriteStream(path.resolve(name));
+          if (!url) throw new Error("No url for downloading, was provided");
+
+          const response = await fetch(url);
 
-          data.on("data", (chunk: any) => progressBar.tick(chunk.length));
-          data.pipe(writer);
-          data.on("end", () => {
-            console.log(
-              decorators.yellow(`Binary "${name}" downloaded to:`),
-              decorators.bright(`"${process.cwd()}".`),
-            );
-            // Add permissions to the binary
-            console.log(decorators.cyan(`Giving permissions to "${name}"`));
-            fs.chmodSync(path.resolve(name), 0o755);
-            resolve();
-          });
+          if (!response.ok)
+            throw Error(response.status + " " + response.statusText);
+
+          const contentLength = response.headers.get(
+            "content-length",
+          ) as string;
+          let loaded = 0;
+
+          const progressBar = multibar.create(parseInt(contentLength, 10), 0);
+
+          const writer = fs.createWriteStream(path.resolve(name));
+          const reader = response.body?.getReader();
+          read();
+
+          async function read() {
+            try {
+              let res = await reader?.read();
+              if (res) {
+                if (res.done) {
+                  writer.close();
+                  resolve();
+                } else if (res.value) {
+                  loaded += res.value.length;
+                  progressBar.increment();
+                  progressBar.update(loaded, {
+                    filename: name,
+                  });
+                  writer.write(res.value);
+                  read();
+                }
+              }
+            } catch (error) {
+              console.error(error);
+              reject(error);
+            }
+          }
```

Comment:
The naming of the function is not clear and it's using many variabe from outer scope, making it hard to extract it and bloating the outer function a bit more, why use recursion in that case ?

Reply:
I think will be more clear to use the same patter that you use in the `utils` package @wirednkod with the `while` loop and `break`ing when is done.

Reply:
Thank you both - Will address it asap